File: crates/juicefront/ui/src/lib/retention.ts
/**
 * Shared retention helpers used both at SSR time (UploadCard.astro) and on
 * the client (enhance-retention.ts). Kept isomorphic: no Node/browser APIs.
 */
import { t, type Locale } from "../i18n";

export const DEFAULT_ALLOWED_TTL_HOURS = [0.5, 1, 6, 12, 24, 72, 168];
export const DEFAULT_TTL_HOURS = 24;

const KNOWN_KEYS: Record<number, `retention.${string}`> = {
  0.5: "retention.30m",
  1: "retention.1h",
  6: "retention.6h",
  12: "retention.12h",
  24: "retention.24h",
  72: "retention.3d",
  168: "retention.7d",
};

/** Localized label for an arbitrary allowed-TTL hour value. */
export function ttlLabel(locale: Locale, hours: number): string {
  const known = KNOWN_KEYS[hours];
  if (known) return t(locale, known);

  const h = Number.isFinite(hours) ? hours : DEFAULT_TTL_HOURS;

  if (h < 1) {
    const mins = Math.round(h * 60);
    return t(locale, mins === 1 ? "retention.minute" : "retention.minutes", {
      count: mins,
    });
  }

  const days = h / 24;
  if (Number.isInteger(days)) {
    return t(locale, days === 1 ? "retention.day" : "retention.days", {
      count: days,
    });
  }

  const num = Number.isInteger(h) ? h : Math.round(h * 100) / 100;
  return t(locale, num === 1 ? "retention.hour" : "retention.hours", {
    count: num,
  });
}

File: crates/juicefront/src/ssr/retention.rs
use crate::i18n;

pub const DEFAULT_ALLOWED_TTL_HOURS: &[f64] = &[0.5, 1.0, 6.0, 12.0, 24.0, 72.0, 168.0];
pub const DEFAULT_TTL_HOURS: f64 = 24.0;

fn known_key(hours: f64) -> Option<&'static str> {
    if hours == 0.5 {
        Some("retention.30m")
    } else if hours == 1.0 {
        Some("retention.1h")
    } else if hours == 6.0 {
        Some("retention.6h")
    } else if hours == 12.0 {
        Some("retention.12h")
    } else if hours == 24.0 {
        Some("retention.24h")
    } else if hours == 72.0 {
        Some("retention.3d")
    } else if hours == 168.0 {
        Some("retention.7d")
    } else {
        None
    }
}

pub fn ttl_label(locale: &str, hours: f64) -> String {
    if let Some(key) = known_key(hours) {
        return i18n::t(locale, key);
    }
    let hours = if hours.is_finite() {
        hours
    } else {
        DEFAULT_TTL_HOURS
    };
    if hours < 1.0 {
        let mins = (hours * 60.0).round() as i64;
        let key = if mins == 1 {
            "retention.minute"
        } else {
            "retention.minutes"
        };
        return i18n::tv(locale, key, "count", &mins.to_string());
    }
    let days = hours / 24.0;
    if days.fract() == 0.0 {
        let days = days as i64;
        let key = if days == 1 {
            "retention.day"
        } else {
            "retention.days"
        };
        return i18n::tv(locale, key, "count", &days.to_string());
    }
    if hours.fract() == 0.0 {
        let whole = hours as i64;
        let key = if whole == 1 {
            "retention.hour"
        } else {
            "retention.hours"
        };
        return i18n::tv(locale, key, "count", &whole.to_string());
    }
    let rounded = (hours * 100.0).round() / 100.0;
    i18n::tv(locale, "retention.hours", "count", &rounded.to_string())
}
